File: dotnet/src/EventGraph/Intelligence.cs
using System.Diagnostics;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace EventGraph;

// ── Provider Interface ──────────────────────────────────────────────────

/// <summary>Extends IIntelligence with metadata about the backing LLM.</summary>
public interface IProvider : IIntelligence
{
    /// <summary>The provider identifier (e.g., "openai", "anthropic", "ollama").</summary>
    string Name { get; }

    /// <summary>The model identifier (e.g., "gpt-4o", "claude-sonnet-4-6").</summary>
    string Model { get; }
}

// ── Config ──────────────────────────────────────────────────────────────

/// <summary>Configuration for creating an IProvider via ProviderFactory.</summary>
public sealed class ProviderConfig
{
    /// <summary>Provider name: "openai", "openai-compatible", "xai", "groq", "together", "ollama", "azure", "claude-cli".</summary>
    public string Provider { get; }

    /// <summary>Model identifier (provider-specific).</summary>
    public string Model { get; }

    /// <summary>API key for authentication. Falls back to environment variables if empty.</summary>
    public string ApiKey { get; init; } = "";

    /// <summary>Overrides the default API endpoint. Useful for proxies, Ollama, Azure, etc.</summary>
    public string BaseUrl { get; init; } = "";

    /// <summary>Caps the response length. Defaults to 1024 if zero.</summary>
    public int MaxTokens { get; init; }

    /// <summary>Controls randomness. Zero means provider default.</summary>
    public double Temperature { get; init; }

    /// <summary>System prompt prepended to every Reason call.</summary>
    public string SystemPrompt { get; init; } = "";

    /// <summary>Create a new ProviderConfig.</summary>
    /// <param name="provider">Provider name (e.g., "openai", "claude-cli").</param>
    /// <param name="model">Model identifier.</param>
    public ProviderConfig(string provider, string model)
    {
        if (string.IsNullOrWhiteSpace(provider))
            throw new EmptyRequiredException("ProviderConfig.Provider");
        if (string.IsNullOrWhiteSpace(model))
            throw new EmptyRequiredException("ProviderConfig.Model");
        Provider = provider;
        Model = model;
    }
}

// ── Factory ─────────────────────────────────────────────────────────────

/// <summary>Creates IProvider instances from ProviderConfig.</summary>
public static class ProviderFactory
{
    /// <summary>Create a provider from the given configuration.</summary>
    /// <exception cref="EventGraphException">If the provider name is unknown or config is invalid.</exception>
    public static IProvider Create(ProviderConfig config)
    {
        var maxTokens = config.MaxTokens > 0 ? config.MaxTokens : 1024;
        var effectiveConfig = new ProviderConfig(config.Provider, config.Model)
        {
            ApiKey = config.ApiKey,
            BaseUrl = config.BaseUrl,
            MaxTokens = maxTokens,
            Temperature = config.Temperature,
            SystemPrompt = config.SystemPrompt,
        };

        return effectiveConfig.Provider.ToLowerInvariant() switch
        {
            "claude-cli" => new ClaudeCliProvider(effectiveConfig),
            "openai-compatible" or "openai" or "xai" or "groq" or "together" or "ollama" or "azure"
                => new OpenAICompatibleProvider(effectiveConfig),
            _ => throw new EventGraphException(
                $"Unknown provider: \"{config.Provider}\" (supported: openai, openai-compatible, xai, groq, together, ollama, azure, claude-cli)"),
        };
    }
}

// ── OpenAI-Compatible Provider ──────────────────────────────────────────

/// <summary>
/// Implements IProvider using the OpenAI Chat Completions API.
/// Compatible with: OpenAI, xAI/Grok, Ollama, Together, Azure OpenAI,
/// Fireworks, Groq, and any OpenAI-compatible endpoint.
/// </summary>
public sealed class OpenAICompatibleProvider : IProvider
{
    private static readonly HttpClient SharedClient = new();

    private readonly HttpClient _client;
    private readonly string _baseUrl;
    private readonly string _apiKey;
    private readonly string _model;
    private readonly int _maxTokens;
    private readonly double _temperature;
    private readonly string _systemPrompt;
    private readonly string _providerName;

    /// <summary>Well-known OpenAI-compatible base URLs.</summary>
    internal const string OpenAIBaseUrl = "https://api.openai.com/v1";
    internal const string XaiBaseUrl = "https://api.x.ai/v1";
    internal const string GroqBaseUrl = "https://api.groq.com/openai/v1";
    internal const string TogetherBaseUrl = "https://api.together.xyz/v1";
    internal const string OllamaBaseUrl = "http://localhost:11434/v1";

    /// <inheritdoc />
    public string Name => _providerName;

    /// <inheritdoc />
    public string Model => _model;

    internal OpenAICompatibleProvider(ProviderConfig config) : this(config, null) { }

    internal OpenAICompatibleProvider(ProviderConfig config, HttpClient? httpClient)
    {
        _client = httpClient ?? SharedClient;
        _model = config.Model;
        _maxTokens = config.MaxTokens;
        _temperature = config.Temperature;
        _systemPrompt = config.SystemPrompt;

        var apiKey = config.ApiKey;
        var baseUrl = config.BaseUrl;
        var providerName = config.Provider.ToLowerInvariant();

        // Map shorthand provider names to default base URLs and env vars.
        switch (providerName)
        {
            case "openai":
                if (string.IsNullOrEmpty(baseUrl)) baseUrl = OpenAIBaseUrl;
                if (string.IsNullOrEmpty(apiKey)) apiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY") ?? "";
                break;
            case "xai":
                if (string.IsNullOrEmpty(baseUrl)) baseUrl = XaiBaseUrl;
                if (string.IsNullOrEmpty(apiKey)) apiKey = Environment.GetEnvironmentVariable("XAI_API_KEY") ?? "";
                break;
            case "groq":
                if (string.IsNullOrEmpty(baseUrl)) baseUrl = GroqBaseUrl;
                if (string.IsNullOrEmpty(apiKey)) apiKey = Environment.GetEnvironmentVariable("GROQ_API_KEY") ?? "";
                break;
            case "together":
                if (string.IsNullOrEmpty(baseUrl)) baseUrl = TogetherBaseUrl;
                if (string.IsNullOrEmpty(apiKey)) apiKey = Environment.GetEnvironmentVariable("TOGETHER_API_KEY") ?? "";
                break;
            case "ollama":
                if (string.IsNullOrEmpty(baseUrl))
                {
                    var host = Environment.GetEnvironmentVariable("OLLAMA_HOST") ?? "http://localhost:11434";
                    baseUrl = host.TrimEnd('/') + "/v1";
                }
                break;
            case "openai-compatible":
                if (string.IsNullOrEmpty(apiKey) && string.IsNullOrEmpty(baseUrl))
                {
                    (apiKey, baseUrl, providerName) = IntelligenceHelpers.DetectProvider();
                }
                break;
        }

        if (string.IsNullOrEmpty(baseUrl))
            baseUrl = OpenAIBaseUrl;

        // Infer provider name from base URL if still generic.
        if (providerName == "openai-compatible")
            providerName = IntelligenceHelpers.InferProviderName(baseUrl);

        _apiKey = apiKey;
        _baseUrl = baseUrl.TrimEnd('/');
        _providerName = providerName;
    }

    /// <inheritdoc />
    public Response Reason(string prompt, IReadOnlyList<Event> history)
    {
        var messages = new List<OpenAIMessage>();

        if (!string.IsNullOrEmpty(_systemPrompt))
        {
            messages.Add(new OpenAIMessage { Role = "system", Content = _systemPrompt });
        }

        var historyText = IntelligenceHelpers.EventsToMessages(history);
        if (!string.IsNullOrEmpty(historyText))
        {
            messages.Add(new OpenAIMessage { Role = "user", Content = historyText });
            messages.Add(new OpenAIMessage { Role = "assistant", Content = "I understand the event history. What would you like me to reason about?" });
        }

        messages.Add(new OpenAIMessage { Role = "user", Content = prompt });

        var reqBody = new OpenAIRequest
        {
            Model = _model,
            Messages = messages,
        };
        if (_maxTokens > 0) reqBody.MaxTokens = _maxTokens;
        if (_temperature > 0) reqBody.Temperature = _temperature;

        var json = JsonSerializer.Serialize(reqBody, OpenAIJsonContext.Default.OpenAIRequest);
        var content = new StringContent(json, Encoding.UTF8, "application/json");

        var request = new HttpRequestMessage(HttpMethod.Post, _baseUrl + "/chat/completions")
        {
            Content = content,
        };
        request.Headers.Add("Accept", "application/json");
        if (!string.IsNullOrEmpty(_apiKey))
        {
            request.Headers.Add("Authorization", "Bearer " + _apiKey);
        }

        HttpResponseMessage response;
        try
        {
            response = _client.Send(request);
        }
        catch (Exception ex)
        {
            throw new EventGraphException($"OpenAI API request failed: {ex.Message}", ex);
        }

        var responseBody = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();

        if (!response.IsSuccessStatusCode)
        {
            // Try to extract error message from JSON body.
            try
            {
                var errResp = JsonSerializer.Deserialize(responseBody, OpenAIJsonContext.Default.OpenAIResponse);
                if (errResp?.Error is not null)
                {
                    throw new EventGraphException($"OpenAI API error (HTTP {(int)response.StatusCode}): {errResp.Error.Message}");
                }
            }
            catch (JsonException) { }
            throw new EventGraphException($"OpenAI API error (HTTP {(int)response.StatusCode}): {responseBody}");
        }

        var result = JsonSerializer.Deserialize(responseBody, OpenAIJsonContext.Default.OpenAIResponse)
            ?? throw new EventGraphException("Failed to parse OpenAI response: null result");

        if (result.Choices is null || result.Choices.Count == 0)
        {
            throw new EventGraphException("OpenAI API returned no choices");
        }

        var responseContent = result.Choices[0].Message?.Content ?? "";
        var tokensUsed = result.Usage?.TotalTokens ?? 0;
        var confidence = IntelligenceHelpers.ParseConfidence(tokensUsed);

        return new Response(responseContent, confidence, tokensUsed);
    }
}

// ── Claude CLI Provider ─────────────────────────────────────────────────

/// <summary>
/// Implements IProvider by shelling out to the claude CLI.
/// Uses whatever authentication Claude Code already has (OAuth, API key, etc.)
/// without requiring separate credentials.
/// </summary>
public sealed class ClaudeCliProvider : IProvider
{
    private readonly string _model;
    private readonly double _maxBudget;
    private readonly string _systemPrompt;
    private readonly string _claudePath;

    /// <inheritdoc />
    public string Name => "claude-cli";

    /// <inheritdoc />
    public string Model => _model;

    internal ClaudeCliProvider(ProviderConfig config)
    {
        _model = string.IsNullOrEmpty(config.Model) ? "sonnet" : config.Model;
        _claudePath = string.IsNullOrEmpty(config.BaseUrl) ? "claude" : config.BaseUrl;
        _systemPrompt = config.SystemPrompt;

        // Repurpose Temperature as max budget hint (dollars). Default $1 per call.
        _maxBudget = config.Temperature > 0 ? config.Temperature : 1.0;
    }

    /// <inheritdoc />
    public Response Reason(string prompt, IReadOnlyList<Event> history)
    {
        var fullPrompt = new StringBuilder();
        var historyText = IntelligenceHelpers.EventsToMessages(history);
        if (!string.IsNullOrEmpty(historyText))
        {
            fullPrompt.Append(historyText);
            fullPrompt.Append("\n---\n\n");
        }
        fullPrompt.Append(prompt);

        var args = new StringBuilder();
        args.Append("-p --output-format json");
        args.Append($" --model {_model}");
        args.Append($" --max-budget-usd {_maxBudget:F2}");
        args.Append(" --no-session-persistence");
        if (!string.IsNullOrEmpty(_systemPrompt))
        {
            args.Append($" --system-prompt \"{_systemPrompt.Replace("\"", "\\\"")}\"");
        }

        var psi = new ProcessStartInfo
        {
            FileName = _claudePath,
            Arguments = args.ToString(),
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        // Remove CLAUDECODE env var to allow nested invocation.
        RemoveEnvironmentVariable(psi, "CLAUDECODE");

        Process proc;
        try
        {
            proc = Process.Start(psi) ?? throw new EventGraphException("Failed to start claude CLI process");
        }
        catch (Exception ex)
        {
            throw new EventGraphException($"Claude CLI not found or failed to start: {ex.Message}", ex);
        }

        proc.StandardInput.Write(fullPrompt.ToString());
        proc.StandardInput.Close();

        var stdout = proc.StandardOutput.ReadToEnd();
        var stderr = proc.StandardError.ReadToEnd();
        proc.WaitForExit();

        if (proc.ExitCode != 0)
        {
            // Check if we got JSON output despite non-zero exit.
            if (!string.IsNullOrEmpty(stdout))
            {
                try
                {
                    var partial = JsonSerializer.Deserialize(stdout, ClaudeCliJsonContext.Default.ClaudeCliResult);
                    if (partial is not null && !string.IsNullOrEmpty(partial.Result))
                    {
                        return ResultToResponse(partial);
                    }
                }
                catch (JsonException) { }
            }
            throw new EventGraphException($"Claude CLI error (exit {proc.ExitCode}): {stderr}");
        }

        ClaudeCliResult result;
        try
        {
            result = JsonSerializer.Deserialize(stdout, ClaudeCliJsonContext.Default.ClaudeCliResult)
                ?? throw new EventGraphException($"Failed to parse claude CLI JSON output: null result");
        }
        catch (JsonException ex)
        {
            throw new EventGraphException($"Failed to parse claude CLI JSON output: {ex.Message}\nraw: {stdout}", ex);
        }

        if (result.IsError)
        {
            throw new EventGraphException($"Claude CLI returned error: {result.Result} (subtype: {result.Subtype})");
        }

        return ResultToResponse(result);
    }

    private static Response ResultToResponse(ClaudeCliResult result)
    {
        var tokensUsed = result.Usage.InputTokens + result.Usage.OutputTokens;
        var confidence = IntelligenceHelpers.ParseConfidence(tokensUsed);
        return new Response(result.Result ?? "", confidence, tokensUsed);
    }

    private static void RemoveEnvironmentVariable(ProcessStartInfo psi, string key)
    {
        // Accessing EnvironmentVariables copies the current process env.
        if (psi.EnvironmentVariables.ContainsKey(key))
        {
            psi.EnvironmentVariables.Remove(key);
        }
    }
}

// ── Helper Methods ──────────────────────────────────────────────────────

/// <summary>Shared helper methods for intelligence providers.</summary>
internal static class IntelligenceHelpers
{
    /// <summary>Convert event history into a simple text context for an LLM.</summary>
    internal static string EventsToMessages(IReadOnlyList<Event> events)
    {
        if (events.Count == 0) return "";

        var sb = new StringBuilder("Event history:\n");
        var limit = Math.Min(events.Count, 20);
        for (var i = 0; i < limit; i++)
        {
            var ev = events[i];
            sb.Append($"- [{ev.Type.Value}] {ev.Id.Value} by {ev.Source.Value}\n");
        }
        if (events.Count > 20)
        {
            sb.Append($"... and {events.Count - 20} more events\n");
        }
        return sb.ToString();
    }

    /// <summary>
    /// Derives a confidence score from token usage. This is a heuristic —
    /// real confidence requires model introspection.
    /// </summary>
    internal static Score ParseConfidence(int tokensUsed)
    {
        // Default to 0.7 — we can't truly measure confidence from outside the model.
        return new Score(0.7);
    }

    /// <summary>
    /// Checks environment variables to auto-detect which OpenAI-compatible provider to use.
    /// </summary>
    internal static (string apiKey, string baseUrl, string name) DetectProvider()
    {
        var xaiKey = Environment.GetEnvironmentVariable("XAI_API_KEY");
        if (!string.IsNullOrEmpty(xaiKey))
            return (xaiKey, OpenAICompatibleProvider.XaiBaseUrl, "xai");

        var groqKey = Environment.GetEnvironmentVariable("GROQ_API_KEY");
        if (!string.IsNullOrEmpty(groqKey))
            return (groqKey, OpenAICompatibleProvider.GroqBaseUrl, "groq");

        var togetherKey = Environment.GetEnvironmentVariable("TOGETHER_API_KEY");
        if (!string.IsNullOrEmpty(togetherKey))
            return (togetherKey, OpenAICompatibleProvider.TogetherBaseUrl, "together");

        var openaiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
        if (!string.IsNullOrEmpty(openaiKey))
            return (openaiKey, OpenAICompatibleProvider.OpenAIBaseUrl, "openai");

        var ollamaHost = Environment.GetEnvironmentVariable("OLLAMA_HOST");
        if (!string.IsNullOrEmpty(ollamaHost))
            return ("", ollamaHost.TrimEnd('/') + "/v1", "ollama");

        return ("", "", "openai-compatible");
    }

    /// <summary>Derives a friendly provider name from the base URL.</summary>
    internal static string InferProviderName(string baseUrl)
    {
        var lower = baseUrl.ToLowerInvariant();
        if (lower.Contains("azure")) return "azure";
        if (lower.Contains("openai.com")) return "openai";
        if (lower.Contains("x.ai")) return "xai";
        if (lower.Contains("groq.com")) return "groq";
        if (lower.Contains("together.xyz")) return "together";
        if (lower.Contains("localhost") || lower.Contains("127.0.0.1")) return "ollama";
        if (lower.Contains("fireworks")) return "fireworks";
        return "openai-compatible";
    }
}

// ── JSON DTOs (OpenAI) ──────────────────────────────────────────────────

internal sealed class OpenAIMessage
{
    [JsonPropertyName("role")]
    public string Role { get; set; } = "";

    [JsonPropertyName("content")]
    public string Content { get; set; } = "";
}

internal sealed class OpenAIRequest
{
    [JsonPropertyName("model")]
    public string Model { get; set; } = "";

    [JsonPropertyName("messages")]
    public List<OpenAIMessage> Messages { get; set; } = new();

    [JsonPropertyName("max_tokens")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public int MaxTokens { get; set; }

    [JsonPropertyName("temperature")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public double Temperature { get; set; }
}

internal sealed class OpenAIChoice
{
    [JsonPropertyName("message")]
    public OpenAIMessage? Message { get; set; }

    [JsonPropertyName("finish_reason")]
    public string? FinishReason { get; set; }
}

internal sealed class OpenAIUsage
{
    [JsonPropertyName("prompt_tokens")]
    public int PromptTokens { get; set; }

    [JsonPropertyName("completion_tokens")]
    public int CompletionTokens { get; set; }

    [JsonPropertyName("total_tokens")]
    public int TotalTokens { get; set; }
}

internal sealed class OpenAIErrorDetail
{
    [JsonPropertyName("message")]
    public string? Message { get; set; }

    [JsonPropertyName("type")]
    public string? Type { get; set; }

    [JsonPropertyName("code")]
    public string? Code { get; set; }
}

internal sealed class OpenAIResponse
{
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("choices")]
    public List<OpenAIChoice>? Choices { get; set; }

    [JsonPropertyName("usage")]
    public OpenAIUsage? Usage { get; set; }

    [JsonPropertyName("error")]
    public OpenAIErrorDetail? Error { get; set; }
}

// ── JSON DTOs (Claude CLI) ──────────────────────────────────────────────

internal sealed class ClaudeCliUsage
{
    [JsonPropertyName("input_tokens")]
    public int InputTokens { get; set; }

    [JsonPropertyName("output_tokens")]
    public int OutputTokens { get; set; }
}

internal sealed class ClaudeCliResult
{
    [JsonPropertyName("type")]
    public string? Type { get; set; }

    [JsonPropertyName("subtype")]
    public string? Subtype { get; set; }

    [JsonPropertyName("is_error")]
    public bool IsError { get; set; }

    [JsonPropertyName("result")]
    public string? Result { get; set; }

    [JsonPropertyName("usage")]
    public ClaudeCliUsage Usage { get; set; } = new();

    [JsonPropertyName("total_cost_usd")]
    public double TotalCostUsd { get; set; }

    [JsonPropertyName("stop_reason")]
    public string? StopReason { get; set; }
}

// ── Source-generated JSON contexts ──────────────────────────────────────

[JsonSerializable(typeof(OpenAIRequest))]
[JsonSerializable(typeof(OpenAIResponse))]
internal partial class OpenAIJsonContext : JsonSerializerContext { }

[JsonSerializable(typeof(ClaudeCliResult))]
internal partial class ClaudeCliJsonContext : JsonSerializerContext { }

// ── Convenience ─────────────────────────────────────────────────────────

/// <summary>Creates a ProviderConfig for the Claude CLI provider.</summary>
public static class ClaudeCliConfig
{
    /// <summary>Create a config for the Claude CLI provider with the given model.</summary>
    /// <param name="model">Model name (defaults to "sonnet" if empty).</param>
    public static ProviderConfig Create(string model = "sonnet")
    {
        if (string.IsNullOrEmpty(model)) model = "sonnet";
        return new ProviderConfig("claude-cli", model);
    }
}

File: ts/src/intelligence.ts
/**
 * Intelligence provider module — factory + implementations for LLM providers.
 * Ported from the Go reference implementation (go/pkg/intelligence/).
 *
 * Supports: OpenAI-compatible (OpenAI, xAI, Groq, Together, Ollama, Azure),
 * and Claude CLI (shells out to `claude -p`).
 *
 * No external runtime dependencies — uses Node's built-in `fetch` and `child_process`.
 */
import { execSync } from "node:child_process";
import { Score } from "./types.js";
import { Response, type Intelligence } from "./decision.js";
import type { Event } from "./event.js";

// ── Provider interface ───────────────────────────────────────────────────

/** Extends Intelligence with metadata about the backing LLM. */
export interface Provider extends Intelligence {
  readonly name: string;
  readonly model: string;
}

// ── Config ───────────────────────────────────────────────────────────────

/** Configuration for creating a Provider via createProvider(). */
export interface Config {
  /** Provider name: "openai-compatible", "openai", "xai", "groq", "together", "ollama", "azure", "claude-cli" */
  provider: string;
  /** Model identifier (provider-specific, e.g. "gpt-4o", "claude-sonnet-4-6"). */
  model: string;
  /** API key for authentication. Falls back to env vars if omitted. */
  apiKey?: string;
  /** Override the default API endpoint. */
  baseUrl?: string;
  /** Maximum response tokens. Defaults to 1024. */
  maxTokens?: number;
  /** Temperature (0-2). Undefined means provider default. */
  temperature?: number;
  /** System prompt prepended to every reason() call. */
  systemPrompt?: string;
}

// ── Well-known base URLs ─────────────────────────────────────────────────

const OPENAI_BASE_URL = "https://api.openai.com/v1";
const XAI_BASE_URL = "https://api.x.ai/v1";
const GROQ_BASE_URL = "https://api.groq.com/openai/v1";
const TOGETHER_BASE_URL = "https://api.together.xyz/v1";
const OLLAMA_BASE_URL = "http://localhost:11434/v1";

// ── Factory ──────────────────────────────────────────────────────────────

/**
 * Creates a Provider from the given Config.
 * Throws on unknown provider or missing required fields.
 */
export function createProvider(config: Config): Provider {
  const maxTokens = config.maxTokens ?? 1024;

  switch (config.provider) {
    case "claude-cli":
      return createClaudeCliProvider(config, maxTokens);

    case "openai-compatible":
    case "openai":
    case "xai":
    case "groq":
    case "together":
    case "ollama":
    case "azure":
      return createOpenAICompatibleProvider(config, maxTokens);

    default:
      throw new Error(
        `Unknown provider: "${config.provider}" (supported: openai-compatible, openai, xai, groq, together, ollama, azure, claude-cli)`,
      );
  }
}

// ── Convenience config ───────────────────────────────────────────────────

/** Creates a Config for the Claude CLI provider. */
export function newClaudeCliConfig(model?: string): Config {
  return {
    provider: "claude-cli",
    model: model || "sonnet",
  };
}

// ── OpenAI-compatible provider ───────────────────────────────────────────

class OpenAICompatibleProvider implements Provider {
  readonly name: string;
  readonly model: string;

  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly maxTokens: number;
  private readonly temperature: number | undefined;
  private readonly systemPrompt: string | undefined;

  constructor(
    name: string,
    model: string,
    baseUrl: string,
    apiKey: string,
    maxTokens: number,
    temperature: number | undefined,
    systemPrompt: string | undefined,
  ) {
    this.name = name;
    this.model = model;
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.apiKey = apiKey;
    this.maxTokens = maxTokens;
    this.temperature = temperature;
    this.systemPrompt = systemPrompt;
    Object.freeze(this);
  }

  async reason(prompt: string, history: Event[]): Promise<Response> {
    const messages: OpenAIMessage[] = [];

    // System prompt.
    if (this.systemPrompt) {
      messages.push({ role: "system", content: this.systemPrompt });
    }

    // Event history as context.
    const historyText = eventsToMessages(history);
    if (historyText) {
      messages.push(
        { role: "user", content: historyText },
        { role: "assistant", content: "I understand the event history. What would you like me to reason about?" },
      );
    }

    // The actual prompt.
    messages.push({ role: "user", content: prompt });

    const reqBody: OpenAIRequest = {
      model: this.model,
      messages,
    };
    if (this.maxTokens > 0) {
      reqBody.max_tokens = this.maxTokens;
    }
    if (this.temperature !== undefined && this.temperature > 0) {
      reqBody.temperature = this.temperature;
    }

    const url = `${this.baseUrl}/chat/completions`;
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }

    const resp = await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(reqBody),
    });

    const body = await resp.text();

    if (!resp.ok) {
      // Try to extract error message from JSON body.
      try {
        const errResp = JSON.parse(body) as OpenAIResponse;
        if (errResp.error?.message) {
          throw new Error(`OpenAI API error (HTTP ${resp.status}): ${errResp.error.message}`);
        }
      } catch (e) {
        if (e instanceof Error && e.message.startsWith("OpenAI API error")) {
          throw e;
        }
      }
      throw new Error(`OpenAI API error (HTTP ${resp.status}): ${body}`);
    }

    const result = JSON.parse(body) as OpenAIResponse;

    if (!result.choices || result.choices.length === 0) {
      throw new Error("OpenAI API returned no choices");
    }

    const content = result.choices[0].message.content;
    const tokensUsed = result.usage?.total_tokens ?? 0;
    const confidence = parseConfidence(tokensUsed);

    return new Response(content, confidence, tokensUsed);
  }
}

function createOpenAICompatibleProvider(config: Config, maxTokens: number): OpenAICompatibleProvider {
  if (!config.model) {
    throw new Error("OpenAI-compatible provider requires a model");
  }

  let apiKey = config.apiKey ?? "";
  let baseUrl = config.baseUrl ?? "";
  let providerName = config.provider;

  // Map shorthand provider names to default base URLs and env vars.
  switch (providerName) {
    case "openai":
      if (!baseUrl) baseUrl = OPENAI_BASE_URL;
      if (!apiKey) apiKey = process.env["OPENAI_API_KEY"] ?? "";
      break;
    case "xai":
      if (!baseUrl) baseUrl = XAI_BASE_URL;
      if (!apiKey) apiKey = process.env["XAI_API_KEY"] ?? "";
      break;
    case "groq":
      if (!baseUrl) baseUrl = GROQ_BASE_URL;
      if (!apiKey) apiKey = process.env["GROQ_API_KEY"] ?? "";
      break;
    case "together":
      if (!baseUrl) baseUrl = TOGETHER_BASE_URL;
      if (!apiKey) apiKey = process.env["TOGETHER_API_KEY"] ?? "";
      break;
    case "ollama":
      if (!baseUrl) {
        const host = process.env["OLLAMA_HOST"] ?? "http://localhost:11434";
        baseUrl = `${host}/v1`;
      }
      break;
    case "azure":
      // Azure requires explicit baseUrl and apiKey.
      break;
    case "openai-compatible":
      if (!apiKey && !baseUrl) {
        const detected = detectProvider();
        apiKey = detected.apiKey;
        baseUrl = detected.baseUrl;
        providerName = detected.name;
      }
      break;
  }

  if (!baseUrl) {
    baseUrl = OPENAI_BASE_URL;
  }

  // Infer provider name from base URL if still generic.
  if (providerName === "openai-compatible") {
    providerName = inferProviderName(baseUrl);
  }

  return new OpenAICompatibleProvider(
    providerName,
    config.model,
    baseUrl,
    apiKey,
    maxTokens,
    config.temperature,
    config.systemPrompt,
  );
}

// ── Claude CLI provider ──────────────────────────────────────────────────

/** JSON output from `claude -p --output-format json`. */
interface ClaudeCliResult {
  type: string;
  subtype: string;
  is_error: boolean;
  result: string;
  usage: {
    input_tokens: number;
    output_tokens: number;
  };
  total_cost_usd: number;
  stop_reason: string;
}

class ClaudeCliProvider implements Provider {
  readonly name = "claude-cli";
  readonly model: string;

  private readonly maxBudget: number;
  private readonly systemPrompt: string | undefined;
  private readonly claudePath: string;

  constructor(
    model: string,
    maxBudget: number,
    systemPrompt: string | undefined,
    claudePath: string,
  ) {
    this.model = model;
    this.maxBudget = maxBudget;
    this.systemPrompt = systemPrompt;
    this.claudePath = claudePath;
    Object.freeze(this);
  }

  reason(prompt: string, history: Event[]): Response {
    // Build the full prompt with history context.
    let fullPrompt = "";
    const historyText = eventsToMessages(history);
    if (historyText) {
      fullPrompt += historyText + "\n---\n\n";
    }
    fullPrompt += prompt;

    // Build command args.
    const args = [
      "-p",
      "--output-format", "json",
      "--model", this.model,
      "--max-budget-usd", this.maxBudget.toFixed(2),
      "--no-session-persistence",
    ];
    if (this.systemPrompt) {
      args.push("--system-prompt", this.systemPrompt);
    }

    const cmd = `${this.claudePath} ${args.map(escapeArg).join(" ")}`;

    // Remove CLAUDECODE env var for nested invocation.
    const env = { ...process.env };
    delete env["CLAUDECODE"];

    let stdout: string;
    try {
      stdout = execSync(cmd, {
        input: fullPrompt,
        env,
        encoding: "utf-8",
        maxBuffer: 10 * 1024 * 1024,
        stdio: ["pipe", "pipe", "pipe"],
      });
    } catch (err: unknown) {
      // Check if we got JSON output despite non-zero exit.
      const execErr = err as { stdout?: string; stderr?: string };
      if (execErr.stdout) {
        try {
          const result = JSON.parse(execErr.stdout) as ClaudeCliResult;
          if (result.result) {
            return this.resultToResponse(result);
          }
        } catch {
          // JSON parse failed, fall through to error.
        }
      }
      const stderr = execErr.stderr ?? "";
      throw new Error(`Claude CLI error: ${(err as Error).message}\nstderr: ${stderr}`);
    }

    const result = JSON.parse(stdout) as ClaudeCliResult;

    if (result.is_error) {
      throw new Error(`Claude CLI returned error: ${result.result} (subtype: ${result.subtype})`);
    }

    return this.resultToResponse(result);
  }

  private resultToResponse(result: ClaudeCliResult): Response {
    const tokensUsed = (result.usage?.input_tokens ?? 0) + (result.usage?.output_tokens ?? 0);
    const confidence = parseConfidence(tokensUsed);
    return new Response(result.result, confidence, tokensUsed);
  }
}

function createClaudeCliProvider(config: Config, _maxTokens: number): ClaudeCliProvider {
  const model = config.model || "sonnet";
  const claudePath = config.baseUrl || "claude";

  // Repurpose temperature as max budget hint (dollars).
  const maxBudget = (config.temperature && config.temperature > 0) ? config.temperature : 1.0;

  return new ClaudeCliProvider(model, maxBudget, config.systemPrompt, claudePath);
}

function escapeArg(arg: string): string {
  // Wrap in quotes if it contains spaces or special characters.
  if (/["\s]/.test(arg)) {
    return `"${arg.replace(/"/g, '\\"')}"`;
  }
  return arg;
}

// ── OpenAI API types ─────────────────────────────────────────────────────

interface OpenAIMessage {
  role: string;
  content: string;
}

interface OpenAIRequest {
  model: string;
  messages: OpenAIMessage[];
  max_tokens?: number;
  temperature?: number;
}

interface OpenAIResponse {
  id?: string;
  choices?: {
    message: OpenAIMessage;
    finish_reason: string;
  }[];
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
  error?: {
    message: string;
    type: string;
    code: string;
  };
}

// ── Helper functions ─────────────────────────────────────────────────────

/**
 * Converts event history into a simple text context
 * suitable for passing as conversation history to an LLM.
 * Caps at 20 events for token efficiency.
 */
export function eventsToMessages(events: Event[]): string {
  if (!events || events.length === 0) {
    return "";
  }

  const lines: string[] = ["Event history:"];
  const limit = Math.min(events.length, 20);

  for (let i = 0; i < limit; i++) {
    const ev = events[i];
    lines.push(`- [${ev.type.value}] ${ev.id.value} by ${ev.source.value}`);
  }

  if (events.length > 20) {
    lines.push(`... and ${events.length - 20} more events`);
  }

  return lines.join("\n") + "\n";
}

/**
 * Extracts a confidence score from token usage.
 * Higher token usage with a stop reason of "end_turn" suggests higher confidence.
 * This is a heuristic — real confidence requires model introspection.
 */
export function parseConfidence(_tokensUsed: number): Score {
  // Default to 0.7 — we can't truly measure confidence from outside the model.
  // Future: use log-probs or model self-assessment.
  return new Score(0.7);
}

/**
 * Auto-detects which OpenAI-compatible provider to use
 * based on environment variables.
 */
export function detectProvider(): { apiKey: string; baseUrl: string; name: string } {
  // Check in order of specificity.
  const xaiKey = process.env["XAI_API_KEY"];
  if (xaiKey) {
    return { apiKey: xaiKey, baseUrl: XAI_BASE_URL, name: "xai" };
  }

  const groqKey = process.env["GROQ_API_KEY"];
  if (groqKey) {
    return { apiKey: groqKey, baseUrl: GROQ_BASE_URL, name: "groq" };
  }

  const togetherKey = process.env["TOGETHER_API_KEY"];
  if (togetherKey) {
    return { apiKey: togetherKey, baseUrl: TOGETHER_BASE_URL, name: "together" };
  }

  const openaiKey = process.env["OPENAI_API_KEY"];
  if (openaiKey) {
    return { apiKey: openaiKey, baseUrl: OPENAI_BASE_URL, name: "openai" };
  }

  // Ollama doesn't need an API key.
  const ollamaHost = process.env["OLLAMA_HOST"];
  if (ollamaHost) {
    return { apiKey: "", baseUrl: `${ollamaHost}/v1`, name: "ollama" };
  }

  return { apiKey: "", baseUrl: "", name: "openai-compatible" };
}

/**
 * Derives a friendly provider name from the base URL.
 */
export function inferProviderName(baseUrl: string): string {
  const lower = baseUrl.toLowerCase();

  if (lower.includes("azure")) return "azure";
  if (lower.includes("openai.com")) return "openai";
  if (lower.includes("x.ai")) return "xai";
  if (lower.includes("groq.com")) return "groq";
  if (lower.includes("together.xyz")) return "together";
  if (lower.includes("localhost") || lower.includes("127.0.0.1")) return "ollama";
  if (lower.includes("fireworks")) return "fireworks";

  return "openai-compatible";
}
